test/unit/plugins/divclass.test.js

```diff
@@ -117,5 +140,48 @@ describe('divclass', () => {
       const output = parser.sourceAndRedactedToHtml(source, redacted);
       expect(output).toEqual("<div class=\"complex-example\"><ul>\n<li>une liste ordonnée</li>\n<li>avec <a href=\"http://example.com\">d'autres contenus rédigés</a></li>\n</ul></div>\n");
     });
+
+    it('can restore content with reordered indexes', () => {
+      const source = "[zero]\n\nzero\n\n[/zero]\n\n[one]\n\none\n\n[/one]";
+      const redacted = "[][1]\n\nun\n\n[/][1]\n\n[][0]\n\nzéro\n\n[/][0]";
+      const output = parser.sourceAndRedactedToHtml(source, redacted);
+      expect(output).toEqual("<div class=\"one\"><p>un</p></div>\n<div class=\"zero\"><p>zéro</p></div>\n");
+    });
+
+    it('can restore content with nesting changes', () => {
+      const source = "[zero]\n\n[one]\n\n\n\n[/one]\n\n[/zero]"
+
+      const unNestedRedaction = "[][0]\n\n\n\n[/][0]\n\n[][1]\n\n\n\n[/][1]";
+      expect(parser.sourceAndRedactedToHtml(source, unNestedRedaction))
+        .toEqual("<div class=\"zero\"></div>\n<div class=\"one\"></div>\n");
+
+      const invertedRedaction = "[][1]\n\n[][0]\n\n\n\n[/][0]\n\n[/][1]";
+      expect(parser.sourceAndRedactedToHtml(source, invertedRedaction))
+        .toEqual("<div class=\"one\"><div class=\"zero\"></div></div>\n");
+
+      const brokenRedaction = "[][0]\n\n[][1]\n\n\n\n[/][0]\n\n[/][1]"
+      expect(parser.sourceAndRedactedToHtml(source, brokenRedaction))
+        .toEqual("<div class=\"zero\"><p><a href=\"\"></a></p></div>\n<p><a href=\"\">/</a></p>\n");
+    });
+
+    it('can restore content that adds extra content', () => {
+      const source = "[first]\n\nFirst\n\n[/first]\n\n[second]\n\nSecond\n\n[/second]";
+
+      const reusedIndex = "[][0]\n\nPremier\n\n[/][0]\n\n[][1]\n\nDeuxième\n\n[/][1]\n\n[][1]\n\nTroisième\n\n[/][1]"
+      expect(parser.sourceAndRedactedToHtml(source, reusedIndex))
+        .toEqual("<div class=\"first\"><p>Premier</p></div>\n<div class=\"second\"><p>Deuxième</p></div>\n<div class=\"second\"><p>Troisième</p></div>\n");
+
+      // in every case, extra redactions default to empty links
+      const extraIndex = "[][0]\n\nPremier\n\n[/][0]\n\n[][1]\n\nDeuxième\n\n[/][1]\n\n[][2]\n\nTroisième\n\n[/][2]"
+      expect(parser.sourceAndRedactedToHtml(source, extraIndex))
+        .toEqual("<div class=\"first\"><p>Premier</p></div>\n<div class=\"second\"><p>Deuxième</p></div>\n<p><a href=\"\"></a></p>\n<p>Troisième</p>\n<p><a href=\"\">/</a></p>\n");
+    });
+
+    it('can NOT restore content if required whitespace is removed', () => {
+      const source = "[clazz]\n\nCat\n\n[/clazz]";
+      const redacted = "[][0]\nChat\n[/][0]";
+      const output = parser.sourceAndRedactedToHtml(source, redacted);
+      expect(output).toEqual("<p><a href=\"\"></a>\nChat\n<a href=\"\">/</a></p>\n");
```

Comment:
This seems appropriate at the parser layer but maybe something we want to aggressively check for at our application layer.

Reply:
Yeah, once I get farther down into integrating this with the i18n sync process I know I'm going to want some instrumentation which can compare redactions provided by the source string with redactions consumed by the translated string.